CRM/Report/Form/Contact/Detail.php

```diff
@@ -975,4 +976,44 @@ public function alterComponentDisplay(&$componentRows) {
     }
   }
 
+  /**
+   * Function to alter from clause by adding checks for permissoned financial type.
+   *
+   * @param string $component
+   *
+   * @return string
+   */
+  public function addPermissionedFromClause($component) {
+    if (is_null($this->financialTypes)) {
+      return NULL;
+    }
+    $financialTypes = 0;
+    if (!empty($this->financialTypes)) {
+      $financialTypes = implode(',', array_keys($this->financialTypes));
+    }
+    $clause = "financial_type_id IN ({$financialTypes})";
+    $whereClause = NULL;
+
+    switch ($component) {
```

Comment:
Switch statements are often a code smell. I think it would be better to have separate functions eg. joinLineItemsFromContribution and call the correct one from the calling code than to pass in component. There are a lot of these switch statements in the Accounts code and in a lot of cases I feel the code would be better with separate functions